### ISSUE TITLE
Test test_threshold_based_eviction is flaky (unexpectedly on-demand downloading remote layer remote)

### ISSUE THREAD

**Issue by bayandin** [neondatabase/neon#4154]
`test_threshold_based_eviction#test_threshold_based_eviction`:

Error message:
```
AssertionError: assert not ['2023-05-04T14:42:57.586772Z  WARN eviction_task{tenant_id=a4f5620478d84e0feff340fb651f8bfb timeline_id=f0d71e20c1b6f...AC0000400C00FFFFFFFF-000000067F000032AC000040140000000008__0000000001696070-0000000003DC76E9 for task kind Eviction\n']
```

Pageserver warning itself (I broke the long line into several smaller ones to make it more readable):
```
2023-05-04T14:42:57.586772Z  
WARN eviction_task{tenant_id=a4f5620478d84e0feff340fb651f8bfb timeline_id=f0d71e20c1b6fcbc94a3d391259a7fc1}:
eviction_iteration{policy_kind="LayerAccessThreshold"}: 
unexpectedly on-demand downloading remote layer remote f0d71e20c1b6fcbc94a3d391259a7fc1/000000067F000032AC0000400C00FFFFFFFF-000000067F000032AC000040140000000008__0000000001696070-0000000003DC76E9 for task kind Eviction
```

See https://neon-github-public-dev.s3.amazonaws.com/reports/pr-4092/debug/4883913165/index.html#suites/3fc871d9ee8127d8501d607e03205abb/2cff07ef9a017a6

**Comment by problame** [comment]
We also have a small number of unexpected on-demand downloads in Prod.
Curiously, with some regularity (ca 1 day apart)

<img width="2552" alt="image" src="https://user-images.githubusercontent.com/956573/236837398-dd479416-69c7-4c72-a22a-6e9b0960432e.png">

**Comment by problame** [comment]
Strong correlation with increase in logical size time spent

<img width="2546" alt="image" src="https://user-images.githubusercontent.com/956573/236840104-c493cd62-eb8b-41be-9b42-50f328b7d98f.png">

**Comment by problame** [comment]
Zooming into one of them, just to demonstrate the correlation

<img width="2202" alt="image" src="https://user-images.githubusercontent.com/956573/236842722-c6b44ca2-1d00-405e-9690-d6ebab1ff51b.png">

**Comment by koivunej** [comment]
From @bayandin's gist:

```
/github/home/.cache/pypoetry/virtualenvs/neon-_pxWMzVK-py3.9/lib/python3.9/site-packages/allure_commons/_allure.py:221: in __call__
    return self._fixture_function(*args, **kwargs)
test_runner/fixtures/neon_fixtures.py:1122: in neon_env_builder
    yield builder
test_runner/fixtures/neon_fixtures.py:850: in __exit__
    self.env.pageserver.assert_no_errors()
test_runner/fixtures/neon_fixtures.py:1691: in assert_no_errors
    assert not errors
E   AssertionError: assert not ['2023-05-16T13:41:03.876833Z  WARN eviction_task{tenant_id=92484835d2e280c5a66c3264c35b1d0f timeline_id=e9239d7e2cdba...AC0000400C0200000000-000000067F000032AC000040140000000022__00000000036FE179-0000000003DCB789 for task kind Eviction\n']
```

- [test_threshold_based_eviction[release-pg14]](https://neon-github-public-dev.s3.amazonaws.com/reports/main/4992306427/index.html#suites/3fc871d9ee8127d8501d607e03205abb/de96eaee52a71ccb/)

**Comment by problame** [comment]
We decided to put this on hold until https://github.com/orgs/neondatabase/projects/38 (async `get_value_reconstruct_data`) is merged and see whether that fixes the issue.

Here's a dump of my internal notes on this issue:

---

- **Hypothesis** the logical size computation does IOs, and it does them inside the tokio executor threads ⇒ we only do 8 IOs at a time, and, block the executor threads while doing int ⇒ other tasks get delayed ⇒ …???… ⇒ imitate accesses come too late ⇒ evictions
    - without validating hypothesis, started attempt to convert logical size computation to async ⇒ needs to convert IOs to async
      - That's project https://github.com/orgs/neondatabase/projects/38 
    - So, do the above in reverse order
    - should try to validate hypothesis first, though, but how? Can we increase the number of tokio threads?
    - Also, isn’t there a semaphore for logical size computation that prevents it from happening concurrently? So, are we actually only blocking 1 tokio thread?
        - Can we get the wall clock time that tokio executor threads (not tokio fs threads) are spending in read / write system calls (to filesystem, not sockets) and compare it against total wall clock time passed in a given interval?
    - Ah hm actually only the synthetic size calculation imitation rate-limits, but, the last_record_lsn size calculation doesn’t. So, 1000 timelines doing it at a same time will contend for the $ncpus executor threads.





- should try to validate hypothesis first, though, but how? Can we increase the number of tokio threads?
 
- Also, isn’t there a semaphore for logical size computation that prevents it from happening concurrently? So, are we actually only blocking 1 tokio thread?
    - Can we get the wall clock time that tokio executor threads (not tokio fs threads) are spending in read / write system calls (to filesystem, not sockets) and compare it against total wall clock time passed in a given interval?
- Ah hm actually only the synthetic size calculation imitation rate-limits, but, the last_record_lsn size calculation doesn’t. So, 1000 timelines doing it at a same time will contend for the $ncpus executor threads.

- Idea for the ??? in above Hyptohesis:
    - timeline imitate accesses never get evicted, that’s for sure, because we get `now` before we imitate the acceses
    - but, there’s also the tenant imitate accesses. we only do them once per tenant, so, tenants with multiple timelines can interleave here
    - what can happen???
    -

**Comment by alexanderlaw** [comment]
I've reproduced the aforementioned warning on a commit from 2023-05-04 with 5 test instances running in parallel, on iterations 1, 4, 3:
```
2024-11-18 11:24:36.515 INFO [neon_fixtures.py:1724] not allowed error: 2024-11-18T08:24:20.330056Z  WARN eviction_task{tenant_id=988eef5ad4157d71954309b1b347c673 timeline_id=33d0e008c5b3f8efcf500bca0020c7c9}:eviction_iteration{policy_kind="LayerAccessThreshold"}:imitate_layer_accesses:imitate_timeline_cached_layer_accesses:calculate_logical_size: unexpectedly on-demand downloading remote layer remote 33d0e008c5b3f8efcf500bca0020c7c9/000000067F000032D60000400D00FFFFFFFF-000000067F000032D6000040150000000005__000000000169B450-0000000003DDE7D1 for task kind Eviction
```

And found that the last commit this reproduced on is 7c076ede~1. On commit 7c076ede, 100 iterations passed successfully for me. So I think this issue should be closed as resolved by https://github.com/neondatabase/neon/pull/4859, otherwise the issue formulation maybe should be changed.

**Comment by problame** [comment]
Thanks @alexanderlaw !
